rustls/src/msgs/codec.rs

```diff
@@ -174,88 +177,149 @@ impl Codec for u64 {
     }
 }
 
-pub fn encode_vec_u8<T: Codec>(bytes: &mut Vec<u8>, items: &[T]) {
-    let len_offset = bytes.len();
-    bytes.push(0);
+/// Represent a Rust `Vec` together with the length prefix type used to encode it in TLS.
+///
+/// This implements various traits to make it easy to use in place of a normal `Vec` and to
+/// easily convert between `Vec` and `TlsVec` at API boundaries.
+#[derive(Clone, Debug)]
+pub struct TlsVec<L, T> {
+    items: Vec<T>,
+    len: PhantomData<L>,
+}
+
+impl<L: LengthPrefix, T: Codec> Codec for TlsVec<L, T>
+where
+    usize: From<L>,
+{
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        let len_offset = bytes.len();
+        bytes.extend(L::ZERO_LEN);
+
+        for i in &self.items {
+            i.encode(bytes);
+        }
 
-    for i in items {
-        i.encode(bytes);
+        let len = bytes.len() - len_offset - L::BYTES;
```

Comment:
This is assuming that the `encode` functions will never shrink the vector, which is true for a valid implementation of `encode`, but not guaranteed by the type system.

For this reason and the previous one, we should define a new trait, something like `trait Extend { type Error; fn extend(&mut self, bytes: &[Self::Item]) -> Result<(), Self::Error> }` and change the definition of the `encode` functions to `encode(&mut dyn Extend<Item=u8>)`.